internal/core/items/component_container.rs

```diff
@@ -51,45 +173,56 @@ impl Item for ComponentContainer {
 
     fn layout_info(
         self: Pin<&Self>,
-        _orientation: Orientation,
-        _window_adapter: &Rc<dyn WindowAdapter>,
+        orientation: Orientation,
+        window_adapter: &Rc<dyn WindowAdapter>,
     ) -> LayoutInfo {
-        LayoutInfo { ..Default::default() }
+        self.ensure_updated(window_adapter.window());
+
+        // Query the component_factory property to force a re-layout when that changes
+        if let Some(rc) = self.component.borrow().clone() {
+            vtable::VRc::borrow_pin(&rc).as_ref().layout_info(orientation)
+        } else {
+            Default::default()
+        }
     }
 
     fn input_event_filter_before_children(
         self: Pin<&Self>,
         _: MouseEvent,
-        _window_adapter: &Rc<dyn WindowAdapter>,
+        window_adapter: &Rc<dyn WindowAdapter>,
         _self_rc: &ItemRc,
     ) -> InputEventFilterResult {
+        self.ensure_updated(window_adapter.window());
         InputEventFilterResult::ForwardAndIgnore
     }
 
     fn input_event(
         self: Pin<&Self>,
         _: MouseEvent,
-        _window_adapter: &Rc<dyn WindowAdapter>,
+        window_adapter: &Rc<dyn WindowAdapter>,
         _self_rc: &ItemRc,
     ) -> InputEventResult {
+        self.ensure_updated(window_adapter.window());
         InputEventResult::EventIgnored
     }
 
     fn focus_event(
         self: Pin<&Self>,
         _: &FocusEvent,
-        _window_adapter: &Rc<dyn WindowAdapter>,
+        window_adapter: &Rc<dyn WindowAdapter>,
         _self_rc: &ItemRc,
     ) -> FocusEventResult {
+        self.ensure_updated(window_adapter.window());
         FocusEventResult::FocusIgnored
     }
 
     fn key_event(
         self: Pin<&Self>,
         _: &KeyEvent,
-        _window_adapter: &Rc<dyn WindowAdapter>,
+        window_adapter: &Rc<dyn WindowAdapter>,
         _self_rc: &ItemRc,
     ) -> KeyEventResult {
+        self.ensure_updated(window_adapter.window());
```

Comment:
Why is this needed for this key event function? It bubbles from inside to outside, no?